website/docs/script/vm-cycle-limits.mdx

```diff
@@ -0,0 +1,163 @@
+---
+id: vm-cycle-limits
+title: "VM Cycle Limits"
+---
+
+import Tabs from "@theme/Tabs";
+import TabItem from "@theme/TabItem";
+import Tooltip from "@components/Tooltip";
+
+# VM Cycle Limits
+
+Cycle limits are used to regulate VM Scripts. <Tooltip>ckb-vm</Tooltip> is a flexible VM that is free to implement many control flow constructs, such as loops or branches. As a result, we will need to enforce certain rules in CKB VM to prevent malicious scripts, such as a script with infinite loops.
+
+We introduce a concept called `cycles`, each VM instruction or syscall will consume some amount of cycles. At consensus level, a scalar `max_block_cycles` field is defined so that the sum of cycles consumed by all scripts in a block cannot exceed this value. Otherwise, the block will be rejected. This way we can guarantee all scripts running in CKB VM will halt, or result in error state.
+
+## max_block_cycles
+
+As mentioned above, a new scalar `max_block_cycles` field is added to chain spec as a consensus rule, it puts a hard limit on how many cycles a block's scripts can consume. No block can consume cycles larger than `max_block_cycles`.
+
+The current `max_block_cycles` in CKB Mainnet `MIRANA` is `3_500_000_000`.
+
+Note there's no limit on the cycles for an individual transaction or a script. As long as the whole block consumes cycles less than `max_block_cycles`, a transaction or a script in that block are free to consume how many cycles they want.
+
+## Cycle Measures
+
+Here we will specify the cycles needed by each CKB VM instructions or syscalls. Note right now in the RFC, we define hard rules for each instruction or syscall here, in future this might be moved into consensus rules so we can change them more easily.
+
+The cycles consumed for each operation are determined based on the following rules:
+
+1. Cycles for RISC-V instructions are determined based on real hardware that implement RISC-V ISA.
+2. Cycles for syscalls are measured based on real runtime performance metrics obtained while benchmarking current CKB implementation.
```

Comment:
```suggestion
Here we will specify the cycles needed by each CKB-VM instruction or syscall.
 
:::note
Currently, we define hard rules for each instruction or syscall in the RFC. In future, this might be moved into consensus rules so we can change them more easily.
:::

The cycles consumed for each operation are determined based on the following rules:

1. Cycles for RISC-V instructions are determined based on real hardware that implements RISC-V ISA.
2. Cycles for syscalls are measured based on real runtime performance metrics obtained while benchmarking current CKB implementation.
```